mlos_bench/mlos_bench/optimizers/mlos_core_optimizer.py

```diff
@@ -128,13 +129,28 @@ def _to_df(self, configs: Sequence[dict]) -> pd.DataFrame:
             A dataframe with past trials data, with missing values imputed.
         """
         df_configs = pd.DataFrame(configs)
-        tunables_names = self._tunables.get_param_values().keys()
+        tunables_names = list(self._tunables.get_param_values().keys())
         missing_cols = set(tunables_names).difference(df_configs.columns)
         for (tunable, _group) in self._tunables:
             if tunable.name in missing_cols:
                 df_configs[tunable.name] = tunable.default
             else:
                 df_configs[tunable.name].fillna(tunable.default, inplace=True)
+            # External data can have incorrect types (e.g., all strings).
+            df_configs[tunable.name] = df_configs[tunable.name].astype(tunable.dtype)
+            # Add columns for tunables with special values.
+            if tunable.special:
+                (special_name, type_name) = special_param_names(tunable.name)
+                tunables_names += [special_name, type_name]
+                is_special = df_configs[tunable.name].apply(tunable.special.__contains__)
+                df_configs[type_name] = "range"
+                df_configs.loc[is_special, type_name] = "special"
+                if tunable.type == "int":
+                    # Make int column NULLABLE:
```

Comment:
Are there new tests for this?  Or did we already have one?

Reply:
we do have tests for the optimzier, but not specifically for the `_to_df()` functionality. I think it is a great idea to add some now. I've opened an issue #636 and will submit a PR soon after we merge this one